### PR TITLE
fix: detect yubikey touches for the new gpg version

### LINKED ISSUE #54: [Feature]: Support new `keyboxd` gpg pubring format
Hey 👋

New `gpg` installations doesn't use `pubring.kbx` format since [`2.4.1`](https://github.com/gpg/gnupg/blob/42ee84197695aca44f5f909a0b1eb59298497da0/README#L131C17-L131C22), and instead use `keyboxd` (aka sqlite db) with the `$HOME/.gnupg/public-keys.d/pubring.db` path.

Would be really awesome if `yubikey-touch-detector` gained support for `keyboxd` format, since newer installations are stuck without gpg detection 🥲

Thanks!

### PR DESCRIPTION
**Pull request by Pablito2020**
The new detection heuristic now checks whether the private key files are opened.

We specifically detect private keys that require a YubiKey touch by filtering out those that have the shadowed-private-key attribute in the S-expression inside the .key file. For more details, refer to the [gnupg documentation that talks about .key format](https://github.com/gpg/gnupg/blob/6737e07a9b04064947ae37abd28b845a09abee22/doc/keyformat.txt#shadowed-private-key-format).

There may be false positives if the user has both a YubiKey and another GPG smartcard device, as the heuristic will detect that gpg opens a private key file. However, this approach is a significant improvement over the current situation, where no devices are detected.

I should mention that I am not an expert in Go (I read just enough to implement this pull request). I originally prototyped this idea in Python and then ported it to Go. Therefore, any feedback on the code would be greatly appreciated, especially if there are any issues with structure or potential bad practices. Let me know if refactoring is needed!

This closes #54

### CODE REVIEW

**Comment by maximbaz** [comment]
Veeery interesting, thanks for finding an approach, as you probably saw in #54 we were struggling to find an event-based way to do the check! I'll try to test tonight!

File: detector/gpg.go [hunk 1]
```diff
@@ -12,17 +12,17 @@ import (
...
-			log.Errorf("Cannot establish a watch on gpg's pubring.kbx file '%v': %v", gpgPubringPath, err)
-			return
-		}
-		log.Debug("GPG watcher is successfully established")
+	  for _, file := range filesToWatch {
+	  	if err := notify.Watch(file, events, notify.InOpen); err != nil {
+	  		log.Errorf("Failed to watch file '%s': %v\n", file, err)
+	  		return;
+	  	}
+	  	log.Debug("GPG watcher is watching '%s'...\n", file)
```
Comment (maximbaz) [inline detector/gpg.go:24]:
```suggestion
	  	log.Debugf("GPG watcher is watching '%s'...\n", file)
```

File: detector/gpg.go [hunk 2]
```diff
@@ -12,17 +12,17 @@ import (
 )
 
 // WatchGPG watches for hints that YubiKey is maybe waiting for a touch on a GPG request
-func WatchGPG(gpgPubringPath string, requestGPGCheck chan bool) {
-	// No need for a buffered channel,
-	// we are interested only in the first event, it's ok to skip all subsequent ones
-	events := make(chan notify.EventInfo)
+func WatchGPG(filesToWatch []string, requestGPGCheck chan bool) {
+	events := make(chan notify.EventInfo, len(filesToWatch))
```
Comment (maximbaz) [inline detector/gpg.go:16]: It's not necessary to add the second parameter here, the comment above is still valid even if we watch multiple files. In the end, we just want to get an event, a signal to tell us to perform an expensive check operation - having one signal is more than enough, it won't really help (and maybe even hurt) if we allow to buffer 5 signals, then having to perform the expensive check 5 times in a row.

Without the second parameter, we stop receiving file events until we execute the expensive operation to check whether yubikey is waiting for a touch, but it's okay, that expensive check will detect the touch regardless of which file originally triggered the event.

So I think let's return the old code snippet here :)

File: detector/gpg.go [hunk 3]
```diff
@@ -12,17 +12,17 @@ import (
...
 
 	initWatcher := func() {
-		if err := notify.Watch(gpgPubringPath, events, notify.InOpen, notify.InDeleteSelf, notify.InMoveSelf); err != nil {
```
Comment (maximbaz) [inline detector/gpg.go:21]: `notify.InDeleteSelf` and `notify.InMoveSelf` here were meant to have resiliency, I don't quite remember the original scenario, but maybe something like if `~/.gnupg` folder was moved, then the app should just re-start gpg detection. Without these, the `default` case below with `GPG received file event '%+v', recreating the watcher.` won't ever be triggered.

Did you remove these on purpose? If not, let's readd them

File: main.go [hunk 4]
```diff
@@ -78,21 +80,66 @@ func main() {
...
-
-			requestGPGCheck := make(chan bool)
-			go detector.CheckGPGOnRequest(requestGPGCheck, notifiers, ctx)
-			go detector.WatchGPG(gpgPubringPath, requestGPGCheck)
-			go detector.WatchSSH(requestGPGCheck, exits)
-		} else {
-			log.Debugf("'%v' could not be found. Disabling GPG and SSH watchers.", gpgPubringPath)
+		var gpgPubringPath = path.Join(gpgme.GetDirInfo("homedir"), "private-keys-v1.d")
+		if _, err := os.Stat(gpgPubringPath); os.IsNotExist(err) {
+			fmt.Printf("Directory '%s' does not exist (you have no private keys).\n", gpgPubringPath)
```
Comment (maximbaz) [inline main.go:85]: Let's use `log.Debugf` everywhere instead of `fmt.Printf`

File: main.go [hunk 5]
```diff
@@ -78,21 +80,66 @@ func main() {
...
-		gpgPubringPath := path.Join(gpgme.GetDirInfo("homedir"), "pubring.kbx")
-		if _, err := os.Stat(gpgPubringPath); err == nil {
-
-			requestGPGCheck := make(chan bool)
-			go detector.CheckGPGOnRequest(requestGPGCheck, notifiers, ctx)
-			go detector.WatchGPG(gpgPubringPath, requestGPGCheck)
-			go detector.WatchSSH(requestGPGCheck, exits)
-		} else {
-			log.Debugf("'%v' could not be found. Disabling GPG and SSH watchers.", gpgPubringPath)
+		var gpgPubringPath = path.Join(gpgme.GetDirInfo("homedir"), "private-keys-v1.d")
```
Comment (maximbaz) [inline main.go:83]: Let's rename to something like `gpgPrivateKeysDirPath`

File: main.go [hunk 6]
```diff
@@ -78,21 +80,66 @@ func main() {
...
+			return err
+		}
+		defer file.Close()
+
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := scanner.Text()
+			if strings.Contains(line, term) {
+				result = append(result, path)
+				break // No need to scan further lines if we already found the string
```
Comment (maximbaz) [inline main.go:128]: you can probably just `return nil` directly from here

File: main.go [hunk 7]
```diff
@@ -78,21 +80,66 @@ func main() {
...
-			go detector.WatchSSH(requestGPGCheck, exits)
-		} else {
-			log.Debugf("'%v' could not be found. Disabling GPG and SSH watchers.", gpgPubringPath)
+		var gpgPubringPath = path.Join(gpgme.GetDirInfo("homedir"), "private-keys-v1.d")
+		if _, err := os.Stat(gpgPubringPath); os.IsNotExist(err) {
+			fmt.Printf("Directory '%s' does not exist (you have no private keys).\n", gpgPubringPath)
+			return
 		}
+		var searchTerm = "shadowed-private-key"
+		var filesToWatch []string
```
Comment (maximbaz) [inline main.go:89]: In Go in most cases you don't need to use `var` if you know the value during declaration, so you can safely remove this line, and line above can be rewritten as `searchTerm := "shadowed-private-key"`

File: main.go [hunk 8]
```diff
@@ -78,21 +80,66 @@ func main() {
...
+		filesToWatch, err := findMatchingFiles(gpgPubringPath, searchTerm)
```
Comment (maximbaz) [inline main.go:90]: minor, but personally I'd rename the function to `findShadowedPrivateKeys`, let it only accept `gpgPrivateKeysDirPath` and it should know itself the search term to look for

**Review by maximbaz — CHANGES_REQUESTED** [review]
Thanks! Some initial feedback - I only managed to test it on the old gpg keyring format (still works!), didn't get to test on the new one, but I believe you tested it anyway, so I have no doubt. Great solution!

Could you also run `go fmt` on the changed files? If not, no worries I'll do it afterwards myself.

File: main.go [hunk 9]
```diff
@@ -78,21 +80,66 @@ func main() {
...
+		if info.IsDir() {
+			return nil
+		}
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		scanner := bufio.NewScanner(file)
```
Comment (maximbaz) [inline main.go:123]: I'd also go for something simpler, the files are small anyway, no need to optimize reading them so much, how does something like this look to you?


```go
		if err != nil || info.IsDir() {
			return err
		}
		data, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		if strings.Contains(string(data), term) {
			result = append(result, path)
		}
		return nil
```

**Comment by maximbaz** [comment]
By the way, I'm happy to take over the PR at any point if it becomes overwhelming, you just let me know at any point of time if you'd like me to merge and cleanup later - you've already done the most important part, figuring out how to support the new gpg format :pray:

File: detector/gpg.go (hunk 2 above)
Comment (Pablito2020) [reply detector/gpg.go]: Good point. We return to the old code snippet.

File: detector/gpg.go (hunk 3 above)
Comment (Pablito2020) [reply detector/gpg.go:21]: No, I just reviewed the Notify documentation, and I initially thought the InOpen documentation was all we needed. I hadn’t considered resiliency—good point! We'll stick with the old one.

File: main.go (hunk 7 above)
Comment (Pablito2020) [reply main.go:88]: "Rookie" mistake. This PR was actually my first time programming in Go! 😆

**Comment by Pablito2020** [comment]
Done!

No worries, it’s not overwhelming, I understand that it was my first time programming in Go, so there was definitely room for improvement!

Tell me if you see something that could be improved.

**Comment by maximbaz** [comment]
It looks great, I just had very minor things that I took the liberty to do myself, I thought it would be more efficient - you can review the last two commits, please feel free to ask if something is not clear. Thanks, going to merge and cut a release!

**Pull request merged**